### PR TITLE
Clarify multi-part AEAD calling sequence in Cipher module

### LINKED ISSUE #4355: Change the cipher.h API to allow delayed output of the last block for AEAD
For some AEAD algorithms, such as GCM or CCM, when using a hardware accelerator, it may not be possible to process partial blocks. This means that if the input to `update()` does not end on a block boundary, it may need to keep a partial block in reserve. The finish function then outputs that last partial block of ciphertext/plaintext. See https://github.com/ARMmbed/mbedtls/issues/4350 for how the problem arose in GCM.

In the `mbedtls/cipher.h` API, for AEAD, the “finish” functions are `mbedtls_cipher_check_tag` and `mbedtls_cipher_write_tag`. Either these functions need to take an extra output buffer, or there needs to be an additional call (to the existing function `mbedtls_cipher_finish`?) to output the last block.

Goal of this task: adjust the `cipher.h` API to support alternative implementations. The implementation in `cipher.c` and the tests may be incompatible with alternative implementations that delay output for partial blocks.

Follow-up: adapt `cipher.c` and the tests https://github.com/ARMmbed/mbedtls/issues/4356

### PR DESCRIPTION
**Pull request by mpg**
Resolves #4355 - turns out a documentation-only change was enough, as the Cipher module already allowed delayed output. Felt the need to clarify the documentation a bit, though, as it didn't say a thing (on way or the other) about whether calling `cipher_finish()` was mandatory for AEAD ciphers as well - now it's explicit.

Considering that it's unclear if we're making a change or just making the documenting the status quo, I'm not sure if we should have a ChangeLog entry and/or migration guide about that.

### CODE REVIEW

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
Looks correct to me apart from a typo. But I think we should consider the fact that it's now mandatory to call finish() for AEAD as an API break. Before, it wasn't documented and doing it for AEAD was always a no-op. So I think this does need a changelog entry and a migration guide.

**Comment by mpg** [comment]
Thanks Gilles. I fixed the typo and added ChangeLog and migration guide entries.

File: ChangeLog.d/cipher-delayed-output.txt [hunk 1]
```diff
@@ -0,0 +1,6 @@
+API changes
+   * For multi-part AEAD operations with the Cipher module, calling
+     mbedtls_cipher_finish() is now mandatory. Previously the documentation
+     was unclear on this point, and this function happened to never do
+     anything with the currently implemented AEADs, so in practice is was
```
Comment (gilles-peskine-arm) [inline ChangeLog.d/cipher-delayed-output.txt:5]: Typo:
```suggestion
     anything with the currently implemented AEADs, so in practice it was
```

File: ChangeLog.d/cipher-delayed-output.txt [hunk 2]
```diff
@@ -0,0 +1,6 @@
+API changes
+   * For multi-part AEAD operations with the Cipher module, calling
```
Comment (gilles-peskine-arm) [inline ChangeLog.d/cipher-delayed-output.txt:2]: Minor: we sometimes write module names in all caps (mostly when they're normally all-caps things, like AES and CTR_DRBG) and sometimes write them in lowercase, but we don't use a leading cap, so
```suggestion
   * For multi-part AEAD operations with the cipher module, calling
```

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
LGTM apart from a typo

File: include/mbedtls/cipher.h [hunk 3]
```diff
@@ -687,7 +696,30 @@ int mbedtls_cipher_set_iv( mbedtls_cipher_context_t *ctx,
...
+ *                4. mbedtls_cipher_finish()
+ *                .
+ *                This sequence can be repeated to encrypt or decrypt multiple
+ *                messages with the same key.
+ *
+ * \note          With AEAD ciphers, the order of calls for each message
+ *                is as follows:
+ *                1. mbedtls_cipher_set_iv() if the mode uses an IV/nonce.
+ *                2. mbedtls_cipher_reset()
+ *                3. mbedtls_cipher_update_ad()
```
Comment (ronald-cron-arm) [inline include/mbedtls/cipher.h:713]: Just wondering here what is the plan around passing additional data in chunks with this cipher interface. No plan to do it, planned for 3.x ?

Comment (mpg) [reply include/mbedtls/cipher.h:713]: I assume it's planned for 3.x, though I guess it's not as urgent as for GCM and CCM (where there's a need for our implementation to match what we require from ALT implementations).

Comment (ronald-cron-arm) [reply include/mbedtls/cipher.h:713]: ok, thanks

**Review by ronald-cron-arm — APPROVED** [review]
Looks good to me.

**Review by gilles-peskine-arm — APPROVED** [review]

**Pull request merged**